### ISSUE TITLE
Chapter 13 Notes, Errata, and Discussion: Fork-Join Parallelism with Futures

### ISSUE THREAD

**Issue by lihaoyi** [handsonscala/handsonscala#13]
Thread to discuss anything and everything about Chapter 13. Important topics will be folded into the issue description.

**Comment by nogurenn** [comment]
For `13.2.2 N Parallel Computations` page 244, the `hash` function still returns a String.
```scala
    hash(p.last)
    // should be wrapped as
    Future{ hash(p.last) }
```

**Comment by lihaoyi** [comment]
@nogurenn the `Future{...}` is around the entire block, on the same line as `val futures =`, so both the `hash` and `println` calls run on the background threadpool

**Comment by nogurenn** [comment]
Oh my bad, you're right

**Comment by implisci** [comment]
The functions in https://github.com/handsonscala/handsonscala/blob/v1/examples/13.7%20-%20ParallelMergeSort/MergeSort.sc  could be in the following order to make it REPL copy and paste friendly.

1. merge
2. mergeSortSequential
3. mergeSortParallel0
4. mergeSortParallel

**Comment by implisci** [comment]
**Use Arrays, mutable Sets, mutable Maps in performance sensitive areas of code.**

Chapter 13 was very interesting and instructive. Took a look at the parallel mergeSort performance on an Array of 10 Million Random Ints. Takes at least 15 seconds on a Dell XPS 13 from 2018. Which is fine, I guess, as it doesn't use any tricks. 

The built in Collections sorted  probably uses many tricks. Array.sorted completes in about 1 second or less on Scala REPL. That's 10x the performance. More interesting is that the .sorted for a Vector with identical content is just under 6 seconds. For performance sensitive steps involving sorting, it is better to convert to Array, then sort, and convert back again to Vector/List or other Seq. In the case of Vector, the sorting now takes 1.4 seconds which is a big improvement on 6 seconds.

For similar reasons, and ones I cannot entirely fathom, .contains in a mutable Set or .get in a mutable Map is much faster (10x in the example below) than .contains in an immutable Set or .get in an immutable Map.  In a tight loop using contains on Sets, performance will be superior with the mutable Set.  Ditto for .get in the case of mutable vs. immutable Maps. Immutable collections like Vector, Set, and Map, are built in such a way to enable reuse of parts of the data structures and save memory while building derivative structures.  This likely imposes a performance penalty for basic operations like sorting, contains and get.  
```scala
// Using Ammonite REPL, Scala REPL runs code a bit faster
val bigArray = Array.tabulate(10000000)(e => scala.util.Random.nextInt)
time {bigArray.sorted}._2.toMillis // 1005 ms
val bigVector = bigArray.toVector
time {bigVector.sorted}._2.toMillis // 7330 ms
time {bigVector.toArray.sorted.toVector}._2.toMillis  // 1316 ms

val bigMutableSet = collection.mutable.Set[Int]().addAll(bigArray) 
time{(-1000000 to 1000000).filter(bigMutableSet.contains(_))}._2.toMillis  // 85 ms
val bigImmutableSet = bigMutableSet.toSet
time{(-1000000 to 1000000).filter(bigMutableSet.contains(_))}._2.toMillis  // 1160 ms
val convertedToMutableSet = collection.mutable.Set[Int]().addAll(bigImmutableSet)
time{(-1000000 to 1000000).filter(convertedToMutableSet.contains(_))}._2.toMillis  // 81 ms

val bigMutableMap = collection.mutable.Map[Int, Int]().addAll(bigArray.indices.zip(bigArray))
time{(0 until 10000000).flatMap(bigMutableMap.get(_))}._2.toMillis   // 619 ms
val bigImmutableMap = bigMutableMap.toMap
time{(0 until 10000000).flatMap(bigImmutableMap.get(_))}._2.toMillis // 8121 ms
val convertedToMutableMap = collection.mutable.Map[Int, Int]().addAll(bigImmutableMap)
time{(0 until 10000000).flatMap(convertedToMutableMap.get(_))}._2.toMillis // 651 ms
```

**Comment by yisraelU** [comment]
I am attempting exercise 13.3.1 and I am getting an error 
`type mismatch;
 found   : requests.Response
 required: ujson.Readable
page <- ujson.read(resp)("query")("pages").obj.values.toSeq
`
Is there supposed to be some implicit conversion ?

**Comment by cdcd88** [comment]
I just finished section 13.5.2 and I don't quite understand `FetchLinksAsync.sc` code

```scala
listenableFut.addListener(() => p.success(listenableFut.get().getResponseBody), null)
val scalaFut: Future[String] = p.future
scalaFut.map{ responseBody =>
  for{
    page <- ujson.read(responseBody)("query")("pages").obj.values.toSeq
    links <- page.obj.get("links").toSeq
    link <- links.arr
  } yield link("title").str
}
```
It explains that this interfaces with callbacks using promise. I don't quite understand this and these are my questions
1. `listenableFut` only handles success results, if it handled error results, would we need to change the following code?
2. Section 13.4.3 mentioned that I could use `foreach` to handle success results. How can I handle error results?

**Comment by branweb1** [comment]
I learned a lot from this chapter. One thing I'm still unsure about thought is the interaction between async operations and threads.

I think I understand the  blocking crawler. We distribute the http call across available threads. They run on those threads. Meanwhile we block on the main thread till they complete. Because `request.get` is blocking, those auxilliary threads are not free while they wait on a response from the server to come back across the network.

What I don't understand is how the async crawler is so much more efficient and is no longer bound by the number of threads. It seems like we still schedule http calls to other threads. I'm guessing because the call this time is asnychronous, the aux thread becomes free right after initializing the call, and when the call completes, the listener is invoked. But what keeps track of the call and invokes the listener? And is it right to assume that the callback still gets scheduled on one of the threads?

**Comment by branweb1** [comment]
> I learned a lot from this chapter. One thing I'm still unsure about thought is the interaction between async operations and threads.
> 
> I think I understand the blocking crawler. We distribute the http call across available threads. They run on those threads. Meanwhile we block on the main thread till they complete. Because `request.get` is blocking, those auxilliary threads are not free while they wait on a response from the server to come back across the network.
> 
> What I don't understand is how the async crawler is so much more efficient and is no longer bound by the number of threads. It seems like we still schedule http calls to other threads. I'm guessing because the call this time is asnychronous, the aux thread becomes free right after initializing the call, and when the call completes, the listener is invoked. But what keeps track of the call and invokes the listener? And is it right to assume that the callback still gets scheduled on one of the threads?

ok after thinking about it a little more, I'm guessing that when a thread runs an async operation, it basically hands it off to the OS and says "tell me when you're done". When the OS does so, the callback attached to that operation runs, either on the same or a different thread. If this is accurate, I still don't get how the OS doesn't get jammed up in the same way as are multi-threaded blocking ap. At some point you run out of things to had the I/O operation off to, so seems like something would need to block at some level.
